rero_ils/modules/loans/logs/statistics.py

```diff
@@ -0,0 +1,152 @@
+#
+# RERO ILS
+# Copyright (C) 2019-2023 RERO
+# Copyright (C) 2019-2023 UCLouvain
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, version 3 of the License.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+"""Methods used to get statistics about `LoanOperationLog`."""
+from copy import deepcopy
+
+from elasticsearch_dsl import A, Q
+from flask import current_app as app
+from invenio_search import RecordsSearch
+
+from rero_ils.modules.loans.logs.api import LoanOperationLog
+
+
+def get_circulation_statistics(filters=None, interval=None, **kwargs):
+    """Get the statistics about circulation operations.
+
+    :param filters: a list of criteria to apply to filter the operation
+        logs index.
+    :param interval: a tuple representing the interval for each step of
+        statistics. Each tuple must provide 2 values :
+          * the interval type : 'calendar_interval' or 'fixed_interval'
+          * the interval value to use corresponding to the type.
+        See official ElasticSearch document for more explanations.
+    :param kwargs: any additional named parameters. Useful key are:
+        * `from` : the transaction lower boundary interval limit (included)
+        * `to`: the transaction upper boundary interval limit (include)
+    """
+    # define default parameters if not specified
+    filters = filters or []
+    filters.append(Q('term', record__type='loan'))
+
+    # build date filter ::
+    #   date filter boundary are build using `from` and `to ` named arguments.
+    #   if one of these argument are equivalent to '*', this boundary will
+    #   be skipped.
+    date_range = {
+        'gte': kwargs.get('from', 'now-7d'),
+        'lte': kwargs.get('to', 'now')
+    }
+    if date_range := {k: v for k, v in date_range.items() if v != '*'}:
+        filters.append(Q('range', date=date_range))
+
+    # build date histogram aggregations
+    interval = interval or ('calendar_interval', '1d')
+    params = {
+        interval[0]: interval[1],  # interval_type: interval_value
+        'format': kwargs.get('format', 'yyyy-MM-dd')
+    }
+    agg = A('date_histogram', field='date', **params)
+    agg.bucket('operation', A('terms', field='loan.trigger'))
+
+    # Build the search and execute it.
+    # Create the stat result data based on search result.
+    query = RecordsSearch(index=LoanOperationLog.index_name) \
+        .query('bool', filter=filters)
+    query.aggs.bucket('operations', agg)
+    query = query[:0]
```

Comment:
Have seen this code before, do we have to use the following lines more often maybe we can have a function for this.